### ISSUE TITLE
ZC158: admin, category-product listing debugs due to function zen_get_discount_calc

### ISSUE THREAD

**Issue by torvista** [zencart/zencart#4287]
ZC158, strict reporting
Admin, search for "test" in product/category listing.
The three categories returned generate three pairs of errors:

> [12-Apr-2021 21:40:04 UTC] Request URI: /zencart-158/admin/index.php?cmd=category_product_listing&search=test, IP address: 127.0.0.1
#1  zen_get_products_base_price() called at [...zencart-158\includes\functions\functions_prices.php:685]
#2  zen_get_discount_calc() called at [...zencart-158\admin\category_product_listing.php:654]
#3  require(...zencart-158\admin\category_product_listing.php) called at [...zencart-158\admin\index.php:11]
--> PHP Notice: Undefined index: products_price in ...zencart-158\includes\functions\functions_prices.php on line 421.

> [12-Apr-2021 21:40:04 UTC] Request URI: /zencart-158/admin/index.php?cmd=category_product_listing&search=test, IP address: 127.0.0.1
> #1  zen_get_products_base_price() called at [...zencart-158\includes\functions\functions_prices.php:685]
> #2  zen_get_discount_calc() called at [...zencart-158\admin\category_product_listing.php:654]
> #3  require(...zencart-158\admin\category_product_listing.php) called at [...zencart-158\admin\index.php:11]
> --> PHP Notice: Undefined index: products_priced_by_attribute in ...zencart-158\includes\functions\functions_prices.php on line 423.

**Comment by torvista** [comment]
I see that apart from this, the table rows are unbalanced on a search, along with other issues which I am looking at on this file....

**Comment by torvista** [comment]
Regarding this issue:
ZC157
The category sale/discount information is returned by
zen_get_products_sale_discount('', $category['categories_id'], true);
from
  function zen_get_products_sale_discount($product_id = false, $categories_id = false, $display_type = false) {

ZC158
This function is deprecated

```
function zen_get_products_sale_discount($product_id = false, $categories_id = false, $display_type = false)
{
    trigger_error('Call to deprecated function zen_get_products_sale_discount. Use zen_get_discount_calc() instead', E_USER_DEPRECATED);
```

But the replacement function does not handle categories:
`function zen_get_discount_calc($product_id, $attribute_id = 0, $attributes_amount = 0, $check_qty = 0)`

This is the function being called in the category/product listing file, hence the errors.

**Comment by mc12345678** [comment]
Suggest changing:
https://github.com/zencart/zencart/blob/35dc5cbc67b99851217a12e580c6130f7b67fd63/admin/category_product_listing.php#L652-L654

to:

````diff
-                <?php if ($show_prod_labels) { ?>
+                <?php if ($show_prod_labels) { 
+                        $product_id = false;
+                        $attribute_id = 1;
+                        $attributes_amount = 0;
+                        if (zen_count_products_in_category($category['categories_id']) > 0) {
+                          $product_id_array = zen_get_categories_products_list($category['categories_id'], false, true, TOPMOST_CATEGORY_PARENT_ID, ' LIMIT 1');
+                          $product_id = $product_id_array[0];
+                          $attribute_id = 0;
+                        }
+                ?>
                  <td class="hidden-sm hidden-xs">&nbsp;</td>
-                <td class="text-right hidden-sm hidden-xs"><?php echo zen_get_discount_calc('', $category['categories_id'], true); ?></td>
+                <td class="text-right hidden-sm hidden-xs"><?php echo zen_get_discount_calc($product_id, $attribute_id, $attributes_amount); ?></td>
````

**Comment by torvista** [comment]
Thanks for the input, I'm not sure if this or the function needs to be modified.

**Comment by proseLA** [comment]
i am not a fan of mc's suggestion here.  there has been a refactor of `zen_get_products_sale_discount` to `zen_get_discount_calc`.  the former method had a default `$product_id` of false.  and somehow that worked for that method.  passing false (which could very easily happen above) looks likely to still result in php notices with the new method.  in addition, i am a fan of only calling the new method if there is a valid `$product_id`.  as to why we are calling this discount_calc method here, i'm still not clear.  

why would we take 1 product from a category list to use for display of a discount?

i am a fan for making code clearer.  which this, for me, does not do.

**Comment by proseLA** [comment]
i think the line:
```
echo zen_get_discount_calc('', $category['categories_id'], true);
```
should be commented out.

**Comment by mc12345678** [comment]
> i am not a fan of mc's suggestion here. there has been a refactor of `zen_get_products_sale_discount` to `zen_get_discount_calc`. the former method had a default `$product_id` of false. and somehow that worked for that method. passing false (which could very easily happen above) looks likely to still result in php notices with the new method. in addition, i am a fan of only calling the new method if there is a valid `$product_id`. as to why we are calling this discount_calc method here, i'm still not clear.
> 
> why would we take 1 product from a category list to use for display of a discount?
> 
> i am a fan for making code clearer. which this, for me, does not do.

So, I haven't a clue why the function `zen_get_products_sale_discount` has been deprecated. It literally only calculates the discount from sale prices where the suggested "replacement" function doesn't provide the same results without either increasing the parameters, stuffing the parameters or some other internal modification. 

As to the relationship of a product to a category, assuming that the product only was on sale (not special, featured or otherwise discounted) then "all" product within that category experience or are subject to the same sale in a default install. That may be why there was a desire to try to display the presence of a sale for a category which would really make it easier to understand what/where sales are located (preferably with some indication of active or otherwise and/or the "type" of sale such as percentage off, actual price or reduction in price).

Mind you, this pick any product within a category for determining the sale "deduction" is central to Zen Cart's sale handling. It is not about discounts, specials or the like, but specifically a sale.  Further this section of code because it is all about the category (which itself has either product or categories, but not both) is guaranteed to deal specifically with sales as compared to the possibility of discounts, specials, or other ways of reducing prices. 

Sure, I'm all for turning that consideration upside down and splitting out sales to support individual product.  Perhaps this discussion is a catalyst for it, but it is not the way that Zen Cart is centrally designed from the front of the catalog to the admin. (I.e., this is a central reason why product can not be in categories that have sub-categories.)
 
So, one of the "issues" (core design concepts) with Zen Cart is that a product is not in a Sale by itself, but in a category. If that is to be modified, then certainly some sort of other operation is necessary. That said, if/with the central concept that a sale applies to the product assigned to a category, then any product that is assigned (by master_categories_id) to that category could be used to identify an aspect of that sale, but the `zen_get_discount_calc` function looks at all aspects of price reduction for the product, not just the sale price. How that function is considered to be the accepted replacement without even providing half of the equivalent functionality is beyond me. It also seems like it was almost a "slap in the face" to push the code specifically to just replace the function name instead of to take further action to either make it functional or for the function to be pulled out from beneath the code... In either way, taking out this additional information because there is (now) nothing built into Zen Cart to support it? Really? It was a feature that was added to be more informative and helpful.  

As for how `zen_get_discount_calc` was used in the proposed code, I had looked at other places where that function was used and at the attempted usage. The other usages didn't "suffer" from the same issue of having to "uniquely" calculate the discounted amount, but instead were "properly" placed as one might expect for identifying the discount.  So, with the idea in mind that `zen_get_discount_calc` just ***had*** to be used, then had to attempt to determine a way to provide the "proposed" result which appeared to be the discount amount that is applicable to the category in question. That this version of the software has been sliced and diced is very surprising. Here we have a deprecation notice to use a function that doesn't provide the functionality of the removed function and this isn't the first such case in this version. The deprecation of `zen_db_output` associated with the admin side's use suggests using `zen_output_string_protected` which doesn't provide the same operation/result. And in trying to cover this topic, I see that the `boxes` class has been deprecated.

If want to revert back to the way it has been by commenting out that line of code, then will also need to go back to the headings for that table and remove the "price/special/sale" list when viewing a category's categories and likely remove other options. Did I make a mistake about which key of the array to use? Sure, but also looked to me like that array was going to be changed to a simple array with keys numbered based on sequence of addition/assignment rather than associated to an identified array key (product). 

Related to the "handling" of `$products_id` being set to false? Yeah there are issues in the "core" for handling those situations and should not reflect on the concept that is proposed here. There are plenty of other outer code calls that reach these inner functions and could/do throw notices.  This set of code prepares the input for those other functions, whether the function(s) properly handle out-of-spec data or not is a separate issue.

**Comment by torvista** [comment]
There seems to be a lot of time and effort expended here from second-guessing the whys and wherefores of code changes and the reasonings behind them. After three months I think some dev input is required to resolve this.

**Comment by scottcwilson** [comment]
Thanks for reminding me - I will take this issue.

**Comment by scottcwilson** [comment]
Fixed by #4552
